model/lists/Service/ListElementsUpdater.php

```diff
@@ -0,0 +1,167 @@
+{
+    private const HUGE_LIST_MIN_ITEMS = 1000;
+    private const HUGE_LIST_MAX_MEMORY_MB = 500;
+
+    /** @var ValueCollectionService */
+    private $valueCollectionService;
+
+    public function __construct(ValueCollectionService $valueCollectionService)
+    {
+        $this->valueCollectionService = $valueCollectionService;
+    }
+
+    public function setListElements(
+        core_kernel_classes_Class $listClass,
+        array $payload
+    ): bool {
+        $clause = $this->getListSearchInput($listClass);
+        $elementsFromPayload = $this->getElementsFromPayload($payload);
+
+        if ($this->isHugeList($clause, $elementsFromPayload)) {
+            $this->raiseMemoryLimit();
+        }
+
+        $valueCollection = $this->valueCollectionService->findAll($clause);
+
+        foreach ($elementsFromPayload as $key => $value) {
+            $this->addElementToCollection($valueCollection, $key, $value);
+        }
+
+        return $this->valueCollectionService->persist($valueCollection);
+    }
+
+    private function addElementToCollection(
+        ValueCollection $valueCollection,
+        $key,
+        $value
+    ): void {
+        $newUriValue = trim($payload["uri_$key"] ?? '');
+        $element = $this->getValueByUriKey($valueCollection, $key);
+
+        if ($element === null) {
+            $valueCollection->addValue(new Value(null, $newUriValue, $value));
+            return;
+        }
+
+        $element->setLabel($value);
+
+        if ($newUriValue) {
+            $element->setUri($newUriValue);
+        }
+    }
+
+    private function getValueByUriKey(ValueCollection $valueCollection, $key): ?Value
+    {
+        $encodedUri = preg_replace('/^list-element_[0-9]+_/', '', $key);
+        $uri = tao_helpers_Uri::decode($encodedUri);
+
+        if (empty($uri)) {
+            return null;
+        }
+
+        return $valueCollection->extractValueByUri($uri);
+    }
+
+    private function getElementsFromPayload(array $payload): array
+    {
+        return array_filter(
+            $payload,
+            function (string $key): bool {
+                return (bool)preg_match('/^list-element_/', $key);
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+    }
+
+    private function getListSearchInput(
+        core_kernel_classes_Class $listClass
+    ): ValueCollectionSearchInput {
+        return new ValueCollectionSearchInput(
+            (new ValueCollectionSearchRequest())->setValueCollectionUri(
+                $listClass->getUri()
+            )
+        );
+    }
+
+    private function isHugeList(
+        ValueCollectionSearchInput $clause,
+        array $listElements
+    ): bool {
+        $itemsToProcess = max(
+            $this->valueCollectionService->count($clause),
+            count($listElements)
+        );
+
+        return $itemsToProcess > self::HUGE_LIST_MIN_ITEMS;
+    }
+
+    private function raiseMemoryLimit(): void
```

Comment:
https://github.com/oat-sa/extension-tao-backoffice/pull/139/files#r841395629

Reply:
Removed in de0da496f4986d330f364da15e621e5789005ee7